File: content/snippets/write-product-descriptions/n2.test.js
/**
 * Ces tests injectent un double local au lieu de charger le point de contrôle.
 *
 * Ce qu'ils prouvent : le dossier arrive au modèle sous la forme de l'affinage,
 * un dossier trop grand est refusé avant toute génération, un appel qui échoue
 * est retenté, une phrase inachevée n'est pas publiée, et une copie qui affirme
 * un attribut absent du dossier est refusée.
 *
 * Ce qu'ils ne prouvent pas : que le modèle écrit bien, ni qu'il invente.
 */
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { FakeSeq2Seq } from '../_harness/fake-model.mjs';
import { FakeLLM } from '../_harness/fake-llm.mjs';
import * as n3 from './n3.js';
import { LocalCopywriter, MAX_CHARACTERS, MIN_CHARACTERS, DescriptionUnavailable, UngroundedDescription, describe } from './n2.js';

const PRODUCT = {
  name: 'Aurore 500',
  category: 'sac à dos',
  material: 'toile recyclée',
  audience: 'les randonneurs',
  features: ['poche pour ordinateur', 'sangle ventrale'],
  colours: ['ardoise', 'sable'],
  warranty: 'deux ans',
};

const VOCABULARY = ['toile recyclée', 'cuir pleine fleur', 'étanche', 'poche pour ordinateur', 'sangle ventrale', 'garanti à vie'];

const COPY =
  'Aurore 500 accompagne les randonneurs à la journée. Sa toile recyclée ' +
  'encaisse les ronces, et sa sangle ventrale reporte la charge sur les hanches.';

const INVENTED =
  'Aurore 500 suit les randonneurs par tous les temps. Sa toile recyclée ' +
  'est entièrement étanche, et sa sangle ventrale reporte la charge.';

/** Un double qui garde aussi les options de génération. */
function recordingModel(answer) {
  const calls = [];
  return { calls, generate: async (source, options) => { calls.push([source, options]); return answer; } };
}

// ---------------------------------------------------------------------------
// Point de rupture
// ---------------------------------------------------------------------------

test('point de rupture : le modèle affirme ce que le dossier ne dit pas', async () => {
  await assert.rejects(() => describe(PRODUCT, new FakeSeq2Seq({}, INVENTED), { vocabulary: VOCABULARY }), (error) => {
    assert.ok(error instanceof UngroundedDescription);
    assert.equal(error.message, 'étanche');
    return true;
  });
  const waterproof = { ...PRODUCT, features: [...PRODUCT.features, 'étanche'] };
  assert.equal(await describe(waterproof, new FakeSeq2Seq({}, INVENTED), { vocabulary: VOCABULARY }), INVENTED);
});

test('point de rupture : un vocabulaire vide laisse passer la même phrase', async () => {
  assert.equal(await describe(PRODUCT, new FakeSeq2Seq({}, INVENTED)), INVENTED);
  assert.equal(await describe(PRODUCT, new FakeSeq2Seq({}, INVENTED), { vocabulary: ['cuir pleine fleur'] }), INVENTED);
});

test('DÉFAUT : un dossier qui nie l’attribut l’ancre quand même', async () => {
  const record = { ...PRODUCT, features: [...PRODUCT.features, 'non étanche'] };
  await assert.rejects(async () => {
    await assert.rejects(() => describe(record, new FakeSeq2Seq({}, INVENTED), { vocabulary: VOCABULARY }), UngroundedDescription);
  }, assert.AssertionError);
});

test('DÉFAUT : un terme de la liste accordé passe le contrôle', async () => {
  await assert.rejects(async () => {
    for (const copy of [
      'Aurore 500 est une lampe solide, garantie à vie par la maison qui la fabrique.',
      'Des coutures étanches et une toile recyclée pour la randonnée du dimanche.',
    ]) {
      await assert.rejects(() => describe(PRODUCT, new FakeSeq2Seq({}, copy), { vocabulary: VOCABULARY }), UngroundedDescription);
    }
  }, assert.AssertionError);
});

// ---------------------------------------------------------------------------
// Docstring et commentaires
// ---------------------------------------------------------------------------

test('écrit la copie que le modèle a rendue', async () => {
  assert.equal(await describe(PRODUCT, new FakeSeq2Seq({}, COPY), { vocabulary: VOCABULARY }), COPY);
});

test('envoie le dossier sous la forme de l’affinage', async () => {
  const model = recordingModel(COPY);
  await describe(PRODUCT, model);
  assert.deepEqual(model.calls[0], [
    'name: Aurore 500 | category: sac à dos | material: toile recyclée | audience: les randonneurs | ' +
      'features: poche pour ordinateur, sangle ventrale | colours: ardoise, sable | warranty: deux ans',
    { max_new_tokens: 90, num_beams: 4 },
  ]);
});

test('ne garde que les phrases que le modèle a finies', async () => {
  const model = new FakeSeq2Seq({}, 'Aurore 500 accompagne les randonneurs à la journée. Sa toile recy');
  assert.equal(await describe(PRODUCT, model), 'Aurore 500 accompagne les randonneurs à la journée.');
});

test('DÉFAUT : un point décimal est pris pour une fin de phrase', async () => {
  const model = new FakeSeq2Seq({}, 'Aurore 500 accompagne les randonneurs à la journée et pèse 1.2 kg avec sa toile recy');
  const published = await describe(PRODUCT, model);
  assert.throws(() => assert.ok(!published.endsWith('1.')), assert.AssertionError);
});

test('un fragment est refusé plutôt que publié', async () => {
  await assert.rejects(() => describe(PRODUCT, new FakeSeq2Seq({}, 'Aurore 500.')), DescriptionUnavailable);
  await assert.rejects(() => describe(PRODUCT, new FakeSeq2Seq({}, 'un sac à dos solide et bien pensé pour la journée')), DescriptionUnavailable);
  const exactly = `${'A'.repeat(MIN_CHARACTERS - 1)}.`;
  assert.equal(await describe(PRODUCT, new FakeSeq2Seq({}, exactly)), exactly);
  await assert.rejects(() => describe(PRODUCT, new FakeSeq2Seq({}, `${'A'.repeat(MIN_CHARACTERS - 2)}.`)), DescriptionUnavailable);
});

test('refuse un dossier trop grand avant de générer quoi que ce soit', async () => {
  const model = new FakeSeq2Seq({}, COPY);
  await assert.rejects(() => describe({ name: 'Aurore 500', features: ['détail interminable '.repeat(40)] }, model), RangeError);
  assert.deepEqual(model.calls, []);
  assert.equal(await describe({ n: 'x'.repeat(MAX_CHARACTERS - 3) }, model), COPY);
  await assert.rejects(() => describe({ n: 'x'.repeat(MAX_CHARACTERS - 2) }, model), RangeError);
});

test('retente un appel qui a échoué', async () => {
  let calls = 0;
  const failingOnce = { async generate() { calls += 1; if (calls === 1) throw new Error('checkpoint not loaded'); return COPY; } };
  assert.equal(await describe(PRODUCT, failingOnce, { attempts: 2 }), COPY);
  assert.equal(calls, 2);
  let always = 0;
  const alwaysFailing = { async generate() { always += 1; throw new Error('checkpoint not loaded'); } };
  await assert.rejects(() => describe(PRODUCT, alwaysFailing), (error) => {
    assert.ok(error instanceof DescriptionUnavailable);
    assert.match(error.message, /checkpoint not loaded/);
    return true;
  });
  assert.equal(always, 2);
});

test('la recherche ignore casse et accents, en mot entier', async () => {
  const shouting = 'Aurore 500 SUIT LES RANDONNEURS. SA TOILE EST ENTIÈREMENT ETANCHE, DIT LA NOTICE.';
  await assert.rejects(() => describe(PRODUCT, new FakeSeq2Seq({}, shouting), { vocabulary: VOCABULARY }), UngroundedDescription);
  const record = { ...PRODUCT, features: ['Étanche'.normalize('NFD')] };
  assert.equal(await describe(record, new FakeSeq2Seq({}, shouting), { vocabulary: VOCABULARY }), shouting);
  assert.ok(await describe(PRODUCT, new FakeSeq2Seq({}, 'Aurore 500 offre une étanchéité soignée aux randonneurs du dimanche.'), { vocabulary: VOCABULARY }));
});

test('le modèle est injecté, et par défaut c’est le vrai', async () => {
  await assert.rejects(() => describe(PRODUCT), (error) => {
    assert.equal(error.code, 'ERR_MODULE_NOT_FOUND');
    assert.match(error.message, /@huggingface\/transformers/);
    return true;
  });
  // La forme du pipeline : `[{ generated_text }]`, lue par LocalCopywriter.
  const seen = [];
  const writer = new LocalCopywriter(async (source, options) => { seen.push(options); return [{ generated_text: COPY }]; });
  assert.equal(await describe(PRODUCT, writer), COPY);
  assert.deepEqual(seen, [{ max_new_tokens: 90, num_beams: 4 }]);
});

test('le même contrôle qu’au niveau N3', async () => {
  const verdict = async (call) => {
    try {
      await call();
      return 'published';
    } catch (error) {
      return `${error.constructor.name}: ${error.message}`;
    }
  };
  for (const copy of [COPY, INVENTED]) {
    const a = await verdict(() => describe(PRODUCT, new FakeSeq2Seq({}, copy), { vocabulary: VOCABULARY }));
    const b = await verdict(() => n3.describe(PRODUCT, new FakeLLM({ response: JSON.stringify({ description: copy }) }), { vocabulary: VOCABULARY }));
    assert.equal(a, b);
  }
});

// ---------------------------------------------------------------------------
// Cas de production
// ---------------------------------------------------------------------------

test('production : dossier vide, zéro essai et attribut nul', async () => {
  let model = recordingModel(COPY);
  assert.equal(await describe({}, model), COPY);
  assert.equal(model.calls[0][0], '');
  await assert.rejects(() => describe(PRODUCT, recordingModel(COPY), { attempts: 0 }), DescriptionUnavailable);
  model = recordingModel(COPY);
  await describe({ name: null }, model);
  assert.equal(model.calls[0][0], 'name: null');
});

test('production : une longue réponse et des espaces en désordre', async () => {
  const longCopy = 'Aurore 500 accompagne les randonneurs.\n\n  '.repeat(2000).trim();
  const start = performance.now();
  const published = await describe(PRODUCT, new FakeSeq2Seq({}, longCopy), { vocabulary: VOCABULARY });
  assert.ok(performance.now() - start < 2000);
  assert.ok(!published.includes('\n') && !published.includes('  '));
});

test('production : une réponse d’un autre type n’est pas publiée', async () => {
  // Python la publie telle quelle (DÉFAUT, n2.test.py) ; ici String() rend « [object Object] », refusé comme fragment.
  await assert.rejects(() => describe(PRODUCT, new FakeSeq2Seq({}, [{ generated_text: 'Aurore 500 accompagne les randonneurs à la journée.' }])), DescriptionUnavailable);
});

test('DÉFAUT : un terme commençant par une ligature n’est jamais trouvé', async () => {
  // `\b` sans drapeau `u` : « œ » n'est pas un caractère de mot, aucune frontière avant lui.
  const copy = 'Sac aux œillets métalliques, pensé pour les randonneurs de la journée.';
  await assert.rejects(async () => {
    await assert.rejects(() => describe(PRODUCT, new FakeSeq2Seq({}, copy), { vocabulary: ['œillets métalliques'] }), UngroundedDescription);
  }, assert.AssertionError);
});

test('DÉFAUT : le plafond compte des unités UTF-16', async () => {
  // 298 emojis : 304 caractères de source en Python, acceptés ; 602 unités UTF-16 ici.
  await assert.rejects(async () => {
    let out;
    try {
      out = await describe({ name: '🙂'.repeat(298) }, new FakeSeq2Seq({}, COPY));
    } catch (error) {
      assert.fail(`${error.name}: ${error.message}`);
    }
    assert.equal(out, COPY);
  }, assert.AssertionError);
});

File: content/snippets/write-product-descriptions/n2.test.py
"""
Ces tests injectent un double local au lieu de charger le point de contrôle.

Ce qu'ils prouvent : le dossier arrive au modèle sous la forme de l'affinage,
un dossier trop grand est refusé avant toute génération, un appel qui échoue
est retenté, une phrase inachevée n'est pas publiée, et une copie qui affirme
un attribut absent du dossier est refusée.

Ce qu'ils ne prouvent pas : que le modèle écrit bien, ni qu'il invente.
"""

import time
import unicodedata

import pytest

from _harness.fake_model import FakeSeq2Seq
import n3
from n2 import BASE_CHECKPOINT, CHECKPOINT, MAX_CHARACTERS, MIN_CHARACTERS, DescriptionUnavailable, UngroundedDescription, describe

# Un produit inventé : aucune marque, aucun catalogue existant.
PRODUCT = {
    "name": "Aurore 500",
    "category": "sac à dos",
    "material": "toile recyclée",
    "audience": "les randonneurs",
    "features": ["poche pour ordinateur", "sangle ventrale"],
    "colours": ["ardoise", "sable"],
    "warranty": "deux ans",
}

# Les mots du catalogue de la boutique : le contrôle d'ancrage vaut exactement cette liste.
VOCABULARY = (
    "toile recyclée",
    "cuir pleine fleur",
    "étanche",
    "poche pour ordinateur",
    "sangle ventrale",
    "garanti à vie",
)

COPY = (
    "Aurore 500 accompagne les randonneurs à la journée. Sa toile recyclée "
    "encaisse les ronces, et sa sangle ventrale reporte la charge sur les hanches."
)

INVENTED = (
    "Aurore 500 suit les randonneurs par tous les temps. Sa toile recyclée "
    "est entièrement étanche, et sa sangle ventrale reporte la charge."
)


class RecordingModel:
    """Un double qui garde aussi les options de génération."""

    def __init__(self, answer):
        self.answer = answer
        self.calls = []

    def generate(self, source, **options):
        self.calls.append((source, options))
        return self.answer


# ---------------------------------------------------------------------------
# Point de rupture
# ---------------------------------------------------------------------------


def test_point_de_rupture_le_modele_affirme_ce_que_le_dossier_ne_dit_pas():
    """
    « Le test fait rendre au double local une phrase juste de ton et de
    grammaire — la toile recyclée du sac y est « entièrement étanche » — pour
    un article dont aucun attribut ne parle d'étanchéité […] L'extrait refuse
    la copie au lieu de la publier » (existant).
    """
    with pytest.raises(UngroundedDescription) as refused:
        describe(PRODUCT, FakeSeq2Seq({}, INVENTED), vocabulary=VOCABULARY)
    assert str(refused.value) == "étanche"
    # Témoin : la même phrase, pour un dossier qui porte l'attribut, passe.
    waterproof = {**PRODUCT, "features": [*PRODUCT["features"], "étanche"]}
    assert describe(waterproof, FakeSeq2Seq({}, INVENTED), vocabulary=VOCABULARY) == INVENTED


def test_point_de_rupture_un_vocabulaire_vide_laisse_passer_la_meme_phrase():
    """« mais seulement pour les termes que vous avez listés : le test suivant montre la même phrase passant intacte avec un vocabulaire vide » (existant)."""
    assert describe(PRODUCT, FakeSeq2Seq({}, INVENTED)) == INVENTED
    assert describe(PRODUCT, FakeSeq2Seq({}, INVENTED), vocabulary=("cuir pleine fleur",)) == INVENTED


def test_defaut_un_dossier_qui_nie_l_attribut_ne_l_ancre_pas():
    record = {**PRODUCT, "features": [*PRODUCT["features"], "non étanche"]}
    with pytest.raises(UngroundedDescription):
        describe(record, FakeSeq2Seq({}, INVENTED), vocabulary=VOCABULARY)


def test_defaut_un_terme_de_la_liste_accorde_est_refuse_aussi():
    for copy in (
        "Aurore 500 est une lampe solide, garantie à vie par la maison qui la fabrique.",
        "Des coutures étanches et une toile recyclée pour la randonnée du dimanche.",
    ):
        with pytest.raises(UngroundedDescription):
            describe(PRODUCT, FakeSeq2Seq({}, copy), vocabulary=VOCABULARY)


# ---------------------------------------------------------------------------
# Docstring et commentaires
# ---------------------------------------------------------------------------


def test_ecrit_la_copie_que_le_modele_a_rendue():
    """Cas nominal (existant)."""
    assert describe(PRODUCT, FakeSeq2Seq({}, COPY), vocabulary=VOCABULARY) == COPY


def test_envoie_le_dossier_sous_la_forme_de_l_affinage():
    """« the source line the model was fine-tuned to read » ; `_source` « one line of « field: value » » (existant, resserré)."""
    model = RecordingModel(COPY)
    describe(PRODUCT, model)
    source, options = model.calls[0]
    assert source == (
        "name: Aurore 500 | category: sac à dos | material: toile recyclée | audience: les randonneurs | "
        "features: poche pour ordinateur, sangle ventrale | colours: ardoise, sable | warranty: deux ans"
    )
    # risks `deterministic: true` : recherche en faisceau, sans tirage.
    assert options == {"max_new_tokens": 90, "num_beams": 4}


def test_ne_garde_que_les_phrases_que_le_modele_a_finies():
    """`_whole_sentences` : « A small model stops when its budget runs out, mid-sentence and sometimes mid-word » (existant)."""
    model = FakeSeq2Seq({}, "Aurore 500 accompagne les randonneurs à la journée. Sa toile recy")
    assert describe(PRODUCT, model) == "Aurore 500 accompagne les randonneurs à la journée."


def test_un_point_decimal_ne_fait_pas_publier_une_copie_tronquee():
    """La copie s'arrête au dernier point, ici celui de « 1.2 » : c'est un fragment, refusé plutôt que publié."""
    model = FakeSeq2Seq({}, "Aurore 500 accompagne les randonneurs à la journée et pèse 1.2 kg avec sa toile recy")
    with pytest.raises(DescriptionUnavailable):
        describe(PRODUCT, model)


def test_un_fragment_est_refuse_plutot_que_publie():
    """`MIN_CHARACTERS` : « Shorter than that, the model handed back a fragment » (existant, complété : limites)."""
    with pytest.raises(DescriptionUnavailable):
        describe(PRODUCT, FakeSeq2Seq({}, "Aurore 500."))
    with pytest.raises(DescriptionUnavailable):
        describe(PRODUCT, FakeSeq2Seq({}, "un sac à dos solide et bien pensé pour la journée"))
    exactly = "A" * (MIN_CHARACTERS - 1) + "."
    assert describe(PRODUCT, FakeSeq2Seq({}, exactly)) == exactly
    with pytest.raises(DescriptionUnavailable):
        describe(PRODUCT, FakeSeq2Seq({}, "A" * (MIN_CHARACTERS - 2) + "."))


def test_refuse_un_dossier_trop_grand_avant_de_generer_quoi_que_ce_soit():
    """« the size cap » (existant, complété : limites)."""
    model = FakeSeq2Seq({}, COPY)
    with pytest.raises(ValueError):
        describe({"name": "Aurore 500", "features": ["détail interminable " * 40]}, model)
    assert model.calls == []
    at_cap = {"n": "x" * (MAX_CHARACTERS - 3)}
    assert describe(at_cap, model) == COPY
    with pytest.raises(ValueError):
        describe({"n": "x" * (MAX_CHARACTERS - 2)}, model)


def test_retente_un_appel_qui_a_echoue():
    """`_generate` : « Retry: on a machine that also serves the shop, the first call fails » (existants, réunis)."""

    class FailingOnce:
        def __init__(self):
            self.calls = 0

        def generate(self, source, **options):
            self.calls += 1
            if self.calls == 1:
                raise MemoryError("checkpoint not loaded")
            return COPY

    model = FailingOnce()
    assert describe(PRODUCT, model, attempts=2) == COPY
    assert model.calls == 2

    class AlwaysFailing:
        calls = 0

        def generate(self, source, **options):
            AlwaysFailing.calls += 1
            raise MemoryError("checkpoint not loaded")

    with pytest.raises(DescriptionUnavailable, match="checkpoint not loaded"):
        describe(PRODUCT, AlwaysFailing())
    assert AlwaysFailing.calls == 2


def test_la_recherche_ignore_casse_et_accents_en_mot_entier():
    """`_says` : « Whole-word search, case and accents set aside » ; `_fold` : « so « Étanche » meets « etanche » »."""
    shouting = "Aurore 500 SUIT LES RANDONNEURS. SA TOILE EST ENTIÈREMENT ETANCHE, DIT LA NOTICE."
    with pytest.raises(UngroundedDescription):
        describe(PRODUCT, FakeSeq2Seq({}, shouting), vocabulary=VOCABULARY)
    record = {**PRODUCT, "features": [unicodedata.normalize("NFD", "Étanche")]}
    assert describe(record, FakeSeq2Seq({}, shouting), vocabulary=VOCABULARY) == shouting
    # Mot entier : « étanchéité » n'est pas « étanche ».
    assert describe(PRODUCT, FakeSeq2Seq({}, "Aurore 500 offre une étanchéité soignée aux randonneurs du dimanche."), vocabulary=VOCABULARY)


def test_le_modele_est_injecte_et_par_defaut_c_est_le_vrai():
    """« `model` is injected so this can be tested without the checkpoint; in production it defaults to the real one above »."""
    with pytest.raises(ModuleNotFoundError, match="transformers"):
        describe(PRODUCT)


def test_le_point_de_controle_de_depart_est_nomme_et_celui_qui_ecrit_est_le_votre():
    """
    docstring : « Start from a model that already writes French: BARThez […]
    What you get back is a checkpoint of your own, which is what `CHECKPOINT`
    points at. » L'extrait ne télécharge rien : il charge vos poids.
    """
    assert BASE_CHECKPOINT == "moussaKam/barthez"
    assert CHECKPOINT.startswith("./")  # un chemin local, pas un dépôt à télécharger
    import inspect

    import n2

    assert inspect.signature(n2.LocalCopywriter.__init__).parameters["checkpoint"].default == CHECKPOINT


def test_le_meme_controle_qu_au_niveau_n3():
    """breaking_point N3 : « Ce qui l'attrape est le contrôle du niveau du dessous » : mêmes verdicts sur les mêmes copies."""
    import json

    from _harness.fake_llm import FakeLLM

    for copy in (COPY, INVENTED):
        def verdict(call):
            try:
                call()
                return "published"
            except Exception as error:  # les deux niveaux ont chacun leur classe
                return f"{type(error).__name__}: {error}"

        n2_verdict = verdict(lambda: describe(PRODUCT, FakeSeq2Seq({}, copy), vocabulary=VOCABULARY))
        n3_verdict = verdict(lambda: n3.describe(PRODUCT, FakeLLM(response=json.dumps({"description": copy})), vocabulary=VOCABULARY))
        assert n2_verdict == n3_verdict


# ---------------------------------------------------------------------------
# Cas de production
# ---------------------------------------------------------------------------


def test_production_dossier_vide_zero_essai_et_attribut_nul():
    """Un dossier sans un attribut ne peut rien donner : il est refusé avant de faire tourner le modèle."""
    model = RecordingModel(COPY)
    with pytest.raises(ValueError, match="empty product record"):
        describe({}, model)
    assert model.calls == []
    # Un attribut à None n'est pas un attribut : le dossier reste vide.
    model = RecordingModel(COPY)
    with pytest.raises(ValueError, match="empty product record"):
        describe({"name": None}, model)
    assert model.calls == []
    with pytest.raises(DescriptionUnavailable):
        describe(PRODUCT, RecordingModel(COPY), attempts=0)


def test_production_une_longue_reponse_et_des_espaces_en_desordre():
    long_copy = ("Aurore 500 accompagne les randonneurs.\n\n  " * 2000).strip()
    start = time.perf_counter()
    published = describe(PRODUCT, FakeSeq2Seq({}, long_copy), vocabulary=VOCABULARY)
    assert time.perf_counter() - start < 2
    assert "\n" not in published and "  " not in published


def test_production_une_reponse_d_un_autre_type_n_est_pas_publiee():
    """`_generate` : une réponse qui n'est pas une chaîne est retentée, puis levée. Même refus en JavaScript."""
    raw = [{"generated_text": "Aurore 500 accompagne les randonneurs à la journée."}]
    with pytest.raises(DescriptionUnavailable):
        describe(PRODUCT, FakeSeq2Seq({}, raw))


def test_production_un_terme_commencant_par_une_ligature_est_trouve():
    """Python : « œillets métalliques » est trouvé (JavaScript le manque, voir n2.test.js)."""
    copy = "Sac aux œillets métalliques, pensé pour les randonneurs de la journée."
    with pytest.raises(UngroundedDescription):
        describe(PRODUCT, FakeSeq2Seq({}, copy), vocabulary=("œillets métalliques",))


def test_production_le_plafond_compte_des_caracteres():
    """Python : 298 emojis font 304 caractères de source, sous le plafond (JavaScript les refuse, voir n2.test.js)."""
    assert describe({"name": "🙂" * 298}, FakeSeq2Seq({}, COPY)) == COPY
